### PR TITLE
CLI - Command Line Interface

### PR DESCRIPTION
**Pull request by marwin1991**
## Summary
- New modules `logchange-commands` and  `logchange-cli`
- Updated JDK version for `1.8` to ` 21`
- Command logic has been moved to the `logchange-commands`
- Created new maven profile for building native images `native`

## TODO
- [x] Fix building
- [x] Tests

### CODE REVIEW

**Comment by marwin1991** [comment]
@witx98 Maybe as you have become prime developer of humble but awesome logchange change commity, and you would like to learn new things this idea will be interesting to you:

create universal CLI which will alow to use this plugin f.e. in swift or C lang projects.

I was thinking about Micronaut framework with GraalVM

https://medium.com/graalvm/cli-applications-with-graalvm-native-image-d629a40aa0be

The idea is to use it also in any system like Linux/Mac/Windows so maybe this CLI should have some installer?

**Comment by witx98** [comment]
To use GraalVM with Micronaut, we should consider adopting a later version of Java, such as Java 21. More details on the benefits and changes can be found here:
- https://micronaut.io/2021/11/23/micronaut-framework-3-2-released/
- https://github.com/micronaut-projects/micronaut-core/issues/7006

We have two options: we could either keep the core modules as they are (Java 8) and update only the CLI module to compile with a newer Java version, or we could upgrade the entire project. However, upgrading would potentially cause compatibility issues for other projects using Logchange.

@marwin1991 
Additionally, if we want this plugin to support Swift and C language projects, we would need a universal way to retrieve the current project version. For example, we could add an argument to the release command. Currently, versioning relies on Maven.

**Comment by marwin1991** [comment]
@witx98 I was thinking about using java8 in all modules besides new module logchange-cli with java 21. Logchange-cli would include core as dependency

**Comment by witx98** [comment]
@marwin1991 Wouldn't it be better to include `logchange-maven-plugin` as a dependency?

**Comment by marwin1991** [comment]
@witx98 I think if something is common, should be moved to some new module or core?

**Comment by witx98** [comment]
@marwin1991  Is this the intended purpose of the `logchange-commands` module, or was it meant for something else? Also, not using `micronaut-parent` requires significantly more configuration - did you consider implementing cli tool as a separate project?

**Comment by marwin1991** [comment]
@witx98 `logchange-cli` was created for implementing there CLI client, `commnads` was extract common parts from `maven-plugin` module. To use micronaut without parent we need:

```
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>io.micronaut</groupId>
                <artifactId>micronaut-bom</artifactId>
                <version>${micronaut.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>
    ```
    
   https://github.com/logchange/logchange/pull/31/files

File: logchange-cli/pom.xml [hunk 1]
```diff
@@ -0,0 +1,309 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <!-- MY I HAVE YOUR ATTENTION PLEASE -->
+    <!-- https://repo1.maven.org/maven2/io/micronaut/micronaut-parent/3.6.1/micronaut-parent-3.6.1.pom -->
```
Comment (marwin1991) [inline logchange-cli/pom.xml:7]: @witx98  ? XD

File: logchange-cli/pom.xml [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        <jib.docker.image>${project.artifactId}</jib.docker.image>
+        <jib.docker.tag>${project.version}</jib.docker.tag>
+    </properties>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>io.micronaut</groupId>
+                <artifactId>micronaut-bom</artifactId>
+                <version>3.10.4</version>
```
Comment (marwin1991) [inline logchange-cli/pom.xml:47]: micronaut.version ?

File: logchange-cli/pom.xml (hunk 1 above)
Comment (witx98) [reply logchange-cli/pom.xml:7]: I'm still cleaning up this PR, but actually, this is your comment 😄 @marwin1991

File: logchange-cli/pom.xml (hunk 2 above)
Comment (witx98) [reply logchange-cli/pom.xml:47]: From what I can see, `micronaut-bom` and `micronaut-core` have separate versioning:  
- https://mvnrepository.com/artifact/io.micronaut/micronaut-bom  
- https://mvnrepository.com/artifact/io.micronaut/micronaut-inject  
@marwin1991

**Comment by janhoy** [comment]
Looking at this tool for potential use with Apache Solr, as we struggle with constant conflicts in `CHANGES.txt` across release branches. However, we use gradle, not maven, so I was hoping to find a CLI version of the tool that we could integrate in our release tooling.

If I understand this PR correctly, it will provide a standalone CLI tool, so if we have a `logchange/` folder with change yml files, we could do something like this:

```bash
./logchange init
./logchange add --empty --fileName=000001-some-name.yml
vi changelog/unreleased/000001-some-name.yml
./logchange lint
./logchange release --version=3.2.1
./logchange generate
cat CHANGELOG.md
```

Also, extracting the logic into independent code will open for a future gradle plugin or similar.

**Comment by witx98** [comment]
Exactly, that's the intention — to enable the use of this tool not only in Maven-based projects but anywhere. We've also considered distributing Docker images to simplify usage across different environments

**Comment by marwin1991** [comment]
@janhoy Hello,
We are working of CLI, and we want to release it with JRelease. 
If you would like to have gradle plugin we can also do that  (before CLI version)

Cc @witx98

**Comment by janhoy** [comment]
Thanks. Looking forward to the CLI tool release, think that is most useful to us right now. JReleaser was new to me, but looks like you can get homebrew installer almost for free -- sweet! Let me know if there are outstanding tasks for which you need  contributions to speed up the next release.

Q: We have been using renovate too but switching to dependabot, is there an integration that lets dependabot generate the yaml file as part of the PR?

**Comment by marwin1991** [comment]
@janhoy we will try to speed up implementation so maybe after Christmas there will be something to test.

Mr. @dangoslen created [tool](
https://github.com/dangoslen/dependabot-changelog-helper) that runs when new PR with given label appears. Maybe @dangoslen will be interested to join logchange family and we will create together gh-action which creates logchange entry based od Dependabot PR :) 🎄

**Comment by marwin1991** [comment]
@witx98 @janhoy 

I am happy to announce that all commands have been implemented and tested in CLI. 

In next merge request we can:
- [provide (--path, -p) option because currently CLI work on current dir (the dir from which we call logchange)](https://github.com/logchange/logchange/issues/373)
- [docker image](https://github.com/logchange/logchange/issues/374) (copy this from valhalla https://github.com/logchange/valhalla)

**Pull request merged**